### PR TITLE
Add connection factory overload for SqlServerStorage

### PR DESCRIPTION
**Pull request by chinwobble**
Solves #1323

### CODE REVIEW

File: src/Hangfire.SqlServer/SqlServerStorage.cs [hunk 1]
```diff
@@ -239,7 +266,9 @@ internal void UseTransaction(DbConnection dedicatedConnection, [InstantHandle] A
 
         internal DbConnection CreateAndOpenConnection()
         {
-            var connection = _existingConnection ?? new SqlConnection(_connectionString);
+            var connection = _existingConnection
```
Comment (odinserj) [inline src/Hangfire.SqlServer/SqlServerStorage.cs]: @chinwobble, thank you for the PR. I think we can simplify this method by introducing the *default connection factory* that returns a new connection based on a connection string. Otherwise we are required to add more tests to see whether `SqlServerStorage` based on a connection factory actually works.

```csharp
() => new SqlConnection(_connectionString);
```

After creating a default connection factory we can simply replace this line with the following one.

```csharp
var connection = _existingConnection ?? _connectionFactory();
```

Comment (odinserj) [inline src/Hangfire.SqlServer/SqlServerStorage.cs]: I think we can simplify this method by introducing the *default connection factory* that returns a new connection based on a connection string. Otherwise we are required to add more tests to see whether `SqlServerStorage` based on a connection factory actually works.

```csharp
() => new SqlConnection(_connectionString);
```

After creating a default connection factory we can simply replace this line with the following one.

```csharp
var connection = _existingConnection ?? _connectionFactory();
```

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs [hunk 2]
```diff
@@ -41,6 +42,15 @@ public void Ctor_CanCreateSqlServerStorage_WithExistingConnection()
             Assert.NotNull(storage);
         }
 
+        [Fact, CleanDatabase]
+        public void Ctor_CanCreateSqlServerStorage_WithConnectionFactory()
+        {
+            Func<DbConnection> connectionFactory = ConnectionUtils.CreateConnection;
+            var storage = new SqlServerStorage(connectionFactory);
+
+            Assert.NotNull(storage);
```
Comment (odinserj) [inline tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: We need to check that the given `connectionFactory` object is used to get a connection. You could use `Mock<DbConnection>` object to check it.

**Comment by chinwobble** [comment]
One thing i Noticed is that in `SqlServerStorage.ToString()` will throw a null reference exception if you pass in a dedicated connection or connectionFactory

File: src/Hangfire.SqlServer/SqlServerStorage.cs (hunk 1 above)
Comment (chinwobble) [reply src/Hangfire.SqlServer/SqlServerStorage.cs]: good suggestion! I've made this change

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs (hunk 2 above)
Comment (chinwobble) [reply tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: yep I've added this change. Let me know if it looks good to you now.

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs [hunk 3]
```diff
@@ -41,6 +43,28 @@ public void Ctor_CanCreateSqlServerStorage_WithExistingConnection()
...
+        }
+
+        [Fact, CleanDatabase]
+        public void CreateAndOpenConnection_UsesConnectionFactory()
+        {
+            var mockConnectionFactory = new Mock<Func<DbConnection>>();
+            var expectedConnection = ConnectionUtils.CreateConnection();
```
Comment (odinserj) [inline tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: You can create `Mock<DbConnection>` instead, pass `() => connection.Object` as a connection factory and check `CreateAndOpenConnection` returns the same `connection.Object` – currently this test anyway fails.

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs (hunk 2 above)
Comment (odinserj) [reply tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: Unfortunately this unit test doesn't test anything, other than *we don't have any exceptions* when creating an instance of the `SqlServerStorage` class – the result of a `new` operator can never be null anyway. 

You can safely remove this test, and optionally add another test that checks the constructor throws `ArgumentNullException`, when connection factory is null. And also another test, that throws the same exception, when options argument is null in factory-based constructor overload.

**Comment by odinserj** [comment]
Also, Travis CI fails, because there were a lot of changes for the .NET Core 1.X build tools. So don't look at it before the `dev` branch is merged into the `master`.

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs (hunk 3 above)
Comment (chinwobble) [reply tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: thanks I got the test working now

File: tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs (hunk 2 above)
Comment (chinwobble) [reply tests/Hangfire.SqlServer.Tests/SqlServerStorageFacts.cs]: done

**Comment by odinserj** [comment]
Thanks for you changes, @chinwobble! I've modified unit test to also check that connection is the same, please see 4dfe0fc commit message. And also reverted commit with `ToString` changes, because it returned parse error for some reason, and we shouldn't establish a connection in the constructor anyway – it doesn't throw anyway in the current implementation, because the body of the `ToString` method is enclosed with a try/catch statement.
